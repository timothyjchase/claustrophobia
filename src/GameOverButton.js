import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Button, Modal, Portal } from 'semantic-ui-react'
import { SCENARIOS } from './config'

class GameOverButton extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      open: props.open || false,
    }
  }

  handleOpen = () => {
    this.setState({ open: true })
  }

  handleClose = () => {
    this.setState({ open: false })
  }

  render() {
    const { as: ButtonComponent, game, result, ...buttonProps } = this.props
    const { open } = this.state

    // use a Portal in order to support a Dropdown.Item trigger component
    return (
      <Portal
        open={open}
        closeOnTriggerClickk={false}
        openOnTriggerClick
        trigger={<ButtonComponent {...buttonProps} />}
        onOpen={this.handleOpen}
        onClose={this.handleClose}
      >
        <Modal open closeOnDimmerClick={false}>
          <Modal.Header>
            {result === 'Defeat' && (
              <span style={{ color: 'red' }}>Maybe next time...</span>
            )}
            {result === 'Victory' && (
              <span style={{ color: 'blue' }}>Congratulations!</span>
            )}
          </Modal.Header>
          <Modal.Content>
            <div>
              <p>
                <strong>Result</strong>: {result}
              </p>
              <p>
                <strong>Scenario</strong>:{' '}
                {(SCENARIOS[game.scenario] || {}).name}
              </p>
              <p>
                <strong>Turn</strong>: {game.turn}
              </p>
              <p>
                <strong>Demons</strong>: {game.demonsInPlay} /{' '}
                {game.demonsAdded}
              </p>
              <p>
                <strong>Troglodytes</strong>: {game.trogsInPlay} /{' '}
                {game.trogsAdded}
              </p>
              <p>
                <strong>Tough Troglodytes</strong>: {game.toughTrogsInPlay} /{' '}
                {game.toughTrogsAdded}
              </p>
              <p>
                <strong>Events</strong>: {game.eventCount}
              </p>
            </div>
          </Modal.Content>
          <Modal.Actions>
            <Button content="Close" onClick={this.handleClose} />
          </Modal.Actions>
        </Modal>
      </Portal>
    )
  }
}

GameOverButton.defaultProps = {
  as: Button,
}

GameOverButton.propTypes = {
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  game: PropTypes.object.isRequired,
  result: PropTypes.string.isRequired,
}

export default GameOverButton
