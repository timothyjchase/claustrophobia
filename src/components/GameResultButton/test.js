import React from 'react'
import GameOverButton from './index'

describe('<GameOverButton />', () => {
  const commonProps = {
    currentState: {
      turn: 1,
      demonsInPlay: 2,
      demonsAdded: 3,
      trogsInPlay: 4,
      trogsAdded: 5,
      toughTrogsInPlay: 6,
      toughTrogsAdded: 7,
      eventCount: 8,
    },
  }
  describe('with result = Victory, scenario', () => {
    const component = shallow(
      <GameOverButton
        {...commonProps}
        result="Victory"
        scenarioName="Test scenario"
      />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
    it('handleOpen changes state to open = true', () => {
      component.instance().handleOpen()
      expect(component.state().open).toEqual(true)
    })
    it('handleClose changes state to open = false', () => {
      component.instance().handleClose()
      expect(component.state().open).toEqual(false)
    })
  })
  describe('with result = Defeat, scenario', () => {
    const component = shallow(
      <GameOverButton
        {...commonProps}
        result="Defeat"
        scenarioName="Test scenario"
      />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('with result = Victory, no scenario', () => {
    const component = shallow(
      <GameOverButton {...commonProps} result="Victory" />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
})
