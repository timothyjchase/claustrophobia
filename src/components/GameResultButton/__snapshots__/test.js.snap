// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameOverButton /> with result = Defeat, scenario should match snapshot 1`] = `
<Portal
  closeOnDocumentClick={true}
  closeOnEscape={true}
  closeOnTriggerClickk={false}
  eventPool="default"
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
  openOnTriggerClick={true}
  trigger={
    <Button
      as="button"
      role="button"
    />
  }
>
  <Modal
    centered={true}
    closeOnDimmerClick={false}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    open={true}
  >
    <ModalHeader>
      <span
        style={
          Object {
            "color": "red",
          }
        }
      >
        Maybe next time...
      </span>
    </ModalHeader>
    <ModalContent>
      <div>
        <p>
          <strong>
            Result: 
          </strong>
          Defeat
        </p>
        <p>
          <strong>
            Scenario: 
          </strong>
          Test scenario
        </p>
        <p>
          <strong>
            Turn: 
          </strong>
          1
        </p>
        <p>
          <strong>
            Demons: 
          </strong>
          2
           / 
          3
        </p>
        <p>
          <strong>
            Troglodytes: 
          </strong>
          4
           / 
          5
        </p>
        <p>
          <strong>
            Tough Troglodytes: 
          </strong>
          6
           / 
          7
        </p>
        <p>
          <strong>
            Events: 
          </strong>
          8
        </p>
      </div>
    </ModalContent>
    <ModalActions>
      <Button
        as="button"
        content="Close"
        onClick={[Function]}
        role="button"
      />
    </ModalActions>
  </Modal>
</Portal>
`;

exports[`<GameOverButton /> with result = Victory, no scenario should match snapshot 1`] = `
<Portal
  closeOnDocumentClick={true}
  closeOnEscape={true}
  closeOnTriggerClickk={false}
  eventPool="default"
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
  openOnTriggerClick={true}
  trigger={
    <Button
      as="button"
      role="button"
    />
  }
>
  <Modal
    centered={true}
    closeOnDimmerClick={false}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    open={true}
  >
    <ModalHeader>
      <span
        style={
          Object {
            "color": "blue",
          }
        }
      >
        Congratulations!
      </span>
    </ModalHeader>
    <ModalContent>
      <div>
        <p>
          <strong>
            Result: 
          </strong>
          Victory
        </p>
        <p>
          <strong>
            Scenario: 
          </strong>
        </p>
        <p>
          <strong>
            Turn: 
          </strong>
          1
        </p>
        <p>
          <strong>
            Demons: 
          </strong>
          2
           / 
          3
        </p>
        <p>
          <strong>
            Troglodytes: 
          </strong>
          4
           / 
          5
        </p>
        <p>
          <strong>
            Tough Troglodytes: 
          </strong>
          6
           / 
          7
        </p>
        <p>
          <strong>
            Events: 
          </strong>
          8
        </p>
      </div>
    </ModalContent>
    <ModalActions>
      <Button
        as="button"
        content="Close"
        onClick={[Function]}
        role="button"
      />
    </ModalActions>
  </Modal>
</Portal>
`;

exports[`<GameOverButton /> with result = Victory, scenario should match snapshot 1`] = `
<Portal
  closeOnDocumentClick={true}
  closeOnEscape={true}
  closeOnTriggerClickk={false}
  eventPool="default"
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
  openOnTriggerClick={true}
  trigger={
    <Button
      as="button"
      role="button"
    />
  }
>
  <Modal
    centered={true}
    closeOnDimmerClick={false}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    open={true}
  >
    <ModalHeader>
      <span
        style={
          Object {
            "color": "blue",
          }
        }
      >
        Congratulations!
      </span>
    </ModalHeader>
    <ModalContent>
      <div>
        <p>
          <strong>
            Result: 
          </strong>
          Victory
        </p>
        <p>
          <strong>
            Scenario: 
          </strong>
          Test scenario
        </p>
        <p>
          <strong>
            Turn: 
          </strong>
          1
        </p>
        <p>
          <strong>
            Demons: 
          </strong>
          2
           / 
          3
        </p>
        <p>
          <strong>
            Troglodytes: 
          </strong>
          4
           / 
          5
        </p>
        <p>
          <strong>
            Tough Troglodytes: 
          </strong>
          6
           / 
          7
        </p>
        <p>
          <strong>
            Events: 
          </strong>
          8
        </p>
      </div>
    </ModalContent>
    <ModalActions>
      <Button
        as="button"
        content="Close"
        onClick={[Function]}
        role="button"
      />
    </ModalActions>
  </Modal>
</Portal>
`;
